lib/stdlib/test/property_test/queue_prop.erl

```diff
@@ -0,0 +1,670 @@
+
+prop_in() ->
+    ?FORALL(
+	L,
+	list(),
+	begin
+	    Q = lists:foldl(
+		fun (I, Acc) ->
+		    queue:in(I, Acc)
+		end,
+		queue:new(),
+		L
+	    ),
+	    equal(L, Q)
+	end
+    ).
+
+prop_in_r() ->
+    common_in_r_cons(in_r).
+
+prop_init() ->
+    common_drop_r_init_liat(init).
+
+prop_is_empty() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    (length(L) =:= 0) =:= queue:is_empty(Q)
+	end
+    ).
+
+prop_join() ->
+    ?FORALL(
+	{{L1, Q1}, {L2, Q2}},
+	{list_queue(), list_queue()},
+	begin
+	    equal(L1 ++ L2, queue:join(Q1, Q2))
+	end
+    ).
+
+prop_last() ->
+    common_get_r_last_daeh(last).
+
+prop_len() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    length(L) =:= queue:len(Q)
+	end
+    ).
+
+prop_liat() ->
+    common_drop_r_init_liat(liat).
+
+prop_member() ->
+    ?FORALL(
+        {X, {L, Q}},
+	{term(), list_queue()},
+	begin
+	    % all members of L are members of Q
+	    lists:all(
+		fun (I) ->
+		    queue:member(I, Q)
+		end,
+		L
+	    )
+	    andalso
+	    % all members of Q are members of L
+	    lists:all(
+		fun (I) ->
+		    lists:member(I, L)
+		end,
+		queue:to_list(Q)
+	    )
+	    andalso
+	    % if X is a member of L, it is also a member of Q,
+	    % and if X is not a member of L, it is also not a
+	    % member of Q
+	    lists:member(X, L) =:= queue:member(X, Q)
+	end
+    ).
+
+prop_out() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    case queue:out(Q) of
+		{{value, I}, Q1} ->
+		    I =:= hd(L) andalso
+		    equal(tl(L), Q1);
+		{empty, Q1} ->
+		    L =:= [] andalso
+		    equal(L, Q1)
+	    end
+	end
+    ).
+
+prop_out_r() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    case queue:out_r(Q) of
+		{{value, I}, Q1} ->
+		    L1 = lists:reverse(L),
+		    I =:= hd(L1) andalso
+		    equal(lists:reverse(tl(L1)), Q1);
+		{empty, Q1} ->
+		    L =:= [] andalso
+		    equal(L, Q1)
+	    end
+	end
+    ).
+
+prop_peek() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    case queue:peek(Q) of
+		{value, I} ->
+		    I =:= hd(L);
+		empty ->
+		    L =:= []
+	    end
+	end
+    ).
+
+prop_peek_r() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    case queue:peek_r(Q) of
+		{value, I} ->
+		    I =:= lists:last(L);
+		empty ->
+		    L =:= []
+	    end
+	end
+    ).
+
+prop_reverse() ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	begin
+	    equal(lists:reverse(L), queue:reverse(Q))
+	end
+    ).
+
+prop_snoc() ->
+    ?FORALL(
+	L,
+	list(),
+	begin
+	    Q = lists:foldl(
+		fun (I, Acc) ->
+		    queue:snoc(Acc, I)
+		end,
+		queue:new(),
+		L
+	    ),
+	    equal(L, Q)
+	end
+    ).
+
+prop_split() ->
+    ?FORALL(
+	{N, {L, Q}},
+	{non_neg_integer(), list_queue()},
+	begin
+	    N1 = N rem (length(L) + 1),
+	    {Q1, Q2} = queue:split(N1, Q),
+	    {L1, L2} = lists:split(N1, L),
+	    equal(L1, Q1) andalso
+	    equal(L2, Q2)
+	end
+    ).
+
+prop_tail() ->
+    common_drop_tail(tail).
+
+% Test sequences of insert and retrieval operations
+prop_ops() ->
```

Comment:
can it be done with "statefull PBT"? https://propertesting.com/book_stateful_properties.html

Reply:
The `prop_ops` property you mean? Maybe, but I doubt it's worth it TBH...

Reply:
Yep, `prop_ops` just looks like a classical example of a statefull test. Not sure if all the PBT frameworks use the same API for statefull tests though.

Reply:
We have some interests in trying out statefull tests in other places but I think here is probably not the use case.